### ISSUE TITLE
Are multithreaded component system updates possible?

### ISSUE THREAD

**Issue by LearnCocos2D** [miguelmartin75/anax#61]
I was wondering in how far anax is multithreading-capable?

1. Can I write component systems whose update(dt) method iterates over the getEntities list on multiple concurrent threads?
2. Would it be possible and safe to update multiple component systems concurrently, each on its own thread?

If you don't have a definite answer I'd be okay with "probably not" or other vague statements. Just so I know whether it's worth investigating or not.

I suppose that at least 1. should be doable, and if so, that would be a major differentiator compared to EntityX and Artemis, as their systems call the update method per entity, passing in a reference to the entity.

**Comment by miguelmartin75** [comment]
>Can I write component systems whose update(dt) method iterates over the getEntities list on multiple concurrent threads?

Yes this should be possible, unless you are modifying components which is shared between threads. e.g. if two systems are running on two separate threads but they require a `Position` component from their entities and one or both modify the `Position` component, then you're bound to be in a data race. You can of course avoid this data race with a mutex, though depending on the situation, it might be more efficient/easier to just run those systems which shared components in the same thread.

>Would it be possible and safe to update multiple component systems concurrently, each on its own thread?

I believe that it is not safe to activate, remove or kill entities from a world in multiple threads. Since I see an obvious data-race with my code (if you look inside the world class), due to me adding the entity to a killed/destroyed/activated/etc. list.

Along with that, even if you don't receive a data-race, there's another issue: if you refresh the world to update every system's entity list, then this refresh/update could occur when another thread is processing it's entities. It would essentially boil down to the following (simplified) situation:

**shared variable:** System's `vector<Entity> entities`

Thread 1 
------------
```c++
// inside your system
for(auto& e : getEntities())
{
    process(e);             
}   
```

Thread 2
----------------                      

 ```c++
system.entities.push_back(e); // which is what happens in world.refresh()                        
```                         

So therefore, you would have to sync up a `world.refresh()` call (i.e. pause all threads until you refresh the world), as otherwise you would run into a data-race. Unless there's a smarter way to avoid the data-race. In fact there may be:

- You could essentially would keep a temporary buffer of entities that should be added/removed to/from the system.
- The next time you call getEntities(), this buffer would be cleared and reflect changes on the system's entity list.
- This temporary buffer will have to be locked, in order to avoid data-races also

If you were to ask for each entity from the world's entity list via a ComponentFilter, then this would be an O(n) search, which is what entityX does, if I'm not mistaken (?). In this case, you would only require a mutex to lock this variable in order to be thread-safe, I think.
